### PR TITLE
Testcase covering child conf mgmt

### PR DESCRIPTION
**Pull request by gligorisaev**
Signed-off-by: Gligor Isaev <glgorisaev@gmail.com>

## Proposed changes

Adding test case for Child device configuration management

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    GET Parent ID                                   #Get the Parent ID from the cloud
+    GET Parent name                                 #Get the Parent name from the cloud
+    GET Child ID                                    #Get the Child ID from the cloud
+    GET Child name                                  #Get the Child name from the cloud
+    # Validate child Name                             #This is to check the existence of the bug: https://github.com/thin-edge/thin-edge.io/issues/1569
+Child device bootstrapping
+    Startup child device                            #Setting up/Bootstraping of a child device
+Snapshot from device
+    Request snapshot from child device              #Using the cloud command Get snapshot from device
+    Child device response on snapshot request       #Child device is sending 'executing' and 'sussessful' MQTT responses
```
Comment (rina23q) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:56]: Always typo :P
```suggestion
    Child device response on snapshot request       #Child device is sending 'executing' and 'successful' MQTT responses
```

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+Child device config update
+    Send configuration to device                    #Using the cloud command Send configuration to device
+    Child device get configuration file             #Child device is sending 'executing' and 'sussessful' MQTT responses
```
Comment (rina23q) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:59]:
```suggestion
    Child device get configuration file             #Child device is sending 'executing' and 'successful' MQTT responses
```

**Review by rina23q — COMMENTED** [review]
Fix typos :P

**Review by rina23q — APPROVED** [review]
Approved. You can use "Squash and Merge"

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+*** Variables ***
+
+${PARENT_IP}             192.168.1.120
+${CHILD_IP}              192.168.1.200
+${HTTP_PORT}             8000
+${USERNAME}              pi
+${PASSWORD}              crypt:LO3wCxZPltyviM8gEyBkRylToqtWm+hvq9mMVEPxtn0BXB65v/5wxUu7EqicpOgGhgNZVgFjY0o=
+${url_tedge}             qaenvironment.eu-latest.cumulocity.com
+${user}                  systest_preparation
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:20]: Just out of curiosity: is there any semantics difference in the variable names in uppercase like `USERNAME` and lowercase like `user` here? Are the uppercase ones used an environmental variables or something like that?

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+${pass}                  crypt:OBusFTXwz00ge67sjHgP8kkH0jLrso7rp6Bp4sHKhwULvkr/nd8WHGezY37/fmMBeNvn+Xxk558glujxb5Pj
+${config}                "files = [\n\t { path = '/home/pi/config1', type = 'config1' },\n ]\n"
+
+${DeviceID}              CH_DEV_CONF_MGMT
+${CHILD}                 sensor1
+${topic_snap}           "tedge/${CHILD}/commands/res/config_snapshot"
+${topic_upd}            "tedge/${CHILD}/commands/res/config_update"
+${topic_restart}        "c8y/s/us/${CHILD}"
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:28]:
```suggestion
${topic_c8y_child}        "c8y/s/us/${CHILD}"
```

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+${payl_restart}         "114,c8y_UploadConfigFile,c8y_DownloadConfigFile"
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:29]:
```suggestion
${payl_supp_ops}         "114,c8y_UploadConfigFile,c8y_DownloadConfigFile"
```
For supported ops.

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    GET Child name                                  #Get the Child name from the cloud
+    # Validate child Name                             #This is to check the existence of the bug: https://github.com/thin-edge/thin-edge.io/issues/1569
+Child device bootstrapping
+    Startup child device                            #Setting up/Bootstraping of a child device
+Snapshot from device
+    Request snapshot from child device              #Using the cloud command Get snapshot from device
+    Child device response on snapshot request       #Child device is sending 'executing' and 'successful' MQTT responses
+Child device config update
+    Send configuration to device                    #Using the cloud command Send configuration to device
+    Child device get configuration file             #Child device is sending 'executing' and 'successful' MQTT responses
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:59]:
```suggestion
    Child device response on update request             #Child device is sending 'executing' and 'successful' MQTT responses
```

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    GET Child ID                                    #Get the Child ID from the cloud
+    GET Child name                                  #Get the Child name from the cloud
+    # Validate child Name                             #This is to check the existence of the bug: https://github.com/thin-edge/thin-edge.io/issues/1569
+Child device bootstrapping
+    Startup child device                            #Setting up/Bootstraping of a child device
+Snapshot from device
+    Request snapshot from child device              #Using the cloud command Get snapshot from device
+    Child device response on snapshot request       #Child device is sending 'executing' and 'successful' MQTT responses
+Child device config update
+    Send configuration to device                    #Using the cloud command Send configuration to device
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:58]:
```suggestion
    Send configuration update to child device                    #Using the cloud command Send configuration to device
```

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    Execute Command    sudo rm c8y-configuration-plugin
+    Close Connection
+
+Child device response on snapshot request    
+    Child Connection
+    ${rc}=    Execute Command    mosquitto_pub -h ${PARENT_IP} -t ${topic_snap} -m ${payl_exec}    return_stdout=False    return_rc=True
+    Should Be Equal    ${rc}    ${0}
+    ${rc}=    Execute Command    curl -X PUT --data-binary @/home/pi/config1 http://${PARENT_IP}:${HTTP_PORT}/tedge/file-transfer/${CHILD}/config_snapshot/config1    return_stdout=False    return_rc=True
+    Should Be Equal    ${rc}    ${0}
+    Sleep    5s
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:171]: Are these sleeps necessary?

File: tests/RobotFramework/config_management/conf_child_mgmt.robot [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    Child Connection
+    ${rc}=    Execute Command    printf ${config} > c8y-configuration-plugin    return_stdout=False    return_rc=True
+    Should Be Equal    ${rc}    ${0}
+    Write    curl -X PUT http://${PARENT_IP}:${HTTP_PORT}/tedge/file-transfer/${CHILD}/c8y-configuration-plugin \\
+    Write   --data-binary @- << EOF
+    Write   files = [
+    Write        { path = '/home/pi/config1', type = 'config1' },
+    Write    ]
+    Write  EOF 
+    Execute Command    sudo apt-get install mosquitto mosquitto-clients -y
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/conf_child_mgmt.robot:137]:
```suggestion
    Execute Command    sudo apt-get install mosquitto-clients -y
```
As the child device only needs the clients and doesn't really need the server.

**Comment by reubenmiller** [comment]
Closing as it is superseded by https://github.com/thin-edge/thin-edge.io/pull/1640

**Pull request closed without merging**